torch/distributions/transforms.py

```diff
@@ -539,3 +541,112 @@ def _call(self, x):
     def _inverse(self, y):
         flat_y = y.contiguous().view((-1,) + y.shape[-2:])
         return torch.stack([self._inverse_on_event(z) for z in flat_y]).view(y.shape)
+
+
+class CatTransform(Transform):
+    """
+    Transform functor that applies a sequence of transforms `tseq`
+    component-wise to each submatrix at `dim` in a way compatible
+    with `torch.cat`.
+    """
+    def __init__(self, tseq, dim=0):
+        assert all(isinstance(t, Transform) for t in tseq)
+        super(CatTransform, self).__init__()
+        self.transforms = list(tseq)
+        self.dim = dim
+    
+    def _call(self, x):
+        assert -x.dim() <= self.dim < x.dim()
+        assert x.size(self.dim) == len(self.transforms)
```

Comment:
This assumes the `.size(self.dim)` of each input is tensor is 1. I think to handle a more general case you could add a `lengths` arg, maybe like
```py
def __init__(self, tseq, dim=0, lengths=None):
    if sizes is None:
        lengths = [1] * len(t_seq)
    assert len(lengths) == len(tseq)
    self.lengths = lengths
    ...

@lazy_property
def length(self):
    return sum(self.lengths)

def _call(self, x):
    ...
    assert x.size(self.dim) == self.length
    yslices = []
    start = 0
    for trans, length in zip(self.transforms, self.lengths):
        xslice = x.narrow(self.dim, start, size)
        yslices.append(trans(xslice))
        start = start + length  # take care to avoid += for jit compatibility
    ...
```